File: recyclerview/src/main/java/com/tunjid/androidbootstrap/recyclerview/AbstractListManagerBuilder.java
package com.tunjid.androidbootstrap.recyclerview;

import android.util.Log;

import com.tunjid.androidbootstrap.functions.BiConsumer;
import com.tunjid.androidbootstrap.functions.Consumer;

import java.util.ArrayList;
import java.util.List;

import androidx.annotation.CallSuper;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.RecyclerView.ViewHolder;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import static com.tunjid.androidbootstrap.recyclerview.ListManager.TAG;

/**
 * Abstract thisInstance class for creating {@link ListManager scrollmanagers}
 * <p></p>
 * The breakdown for the generic types are as follows:
 * <p></p>
 * B: The implicit type of the Builder. This is only necessary to make the return type for each
 * thisInstance method be the type of any inheriting class, and not this base class. Otherwise inheritors of
 * this subclass will need to override each method here to return their custom thisInstance type to
 * maintain the fluency of the API which is fairly tedious.
 * <p></p>
 * S: The type of the {@link ListManager} to be built.
 * <p></p>
 * VH: The {@link ViewHolder} type in the {@link RecyclerView}
 * <p></p>
 * T: The type bound in the {@link ListPlaceholder}
 */
@SuppressWarnings({"unused", "WeakerAccess"})
public abstract class AbstractListManagerBuilder<
        B extends AbstractListManagerBuilder<B, S, VH, T>,
        S extends ListManager<VH, T>,
        VH extends ViewHolder,
        T> {

    private static final int LINEAR_LAYOUT_MANAGER = 1;
    private static final int GRID_LAYOUT_MANAGER = 2;
    private static final int STAGGERED_GRID_LAYOUT_MANAGER = 3;

    protected int spanCount;
    protected int layoutManagerType;
    protected int endlessScrollVisibleThreshold;
    protected boolean hasFixedSize;

    protected ListPlaceholder<T> placeholder;
    protected SwipeRefreshLayout refreshLayout;

    protected RecyclerView recyclerView;
    protected RecyclerView.Adapter<? extends VH> adapter;
    protected RecyclerView.LayoutManager customLayoutManager;

    protected SwipeDragOptions<VH> swipeDragOptions;
    protected RecyclerView.RecycledViewPool recycledViewPool;

    protected Consumer<Integer> endlessScrollConsumer;
    protected Consumer<RecyclerView.LayoutManager> layoutManagerConsumer;
    protected Consumer<IndexOutOfBoundsException> handler;

    protected List<RecyclerView.ItemDecoration> itemDecorations = new ArrayList<>();
    protected List<Consumer<Integer>> stateConsumers = new ArrayList<>();
    protected List<BiConsumer<Integer, Integer>> displacementConsumers = new ArrayList<>();

    protected final B thisInstance;

    @SuppressWarnings("unchecked")
    public AbstractListManagerBuilder() { thisInstance = (B) this;}

    public final B setHasFixedSize() {
        this.hasFixedSize = true;
        return thisInstance;
    }

    public final B withRecyclerView(@NonNull RecyclerView recyclerView) {
        this.recyclerView = recyclerView;
        return thisInstance;
    }

    public final B withAdapter(@NonNull RecyclerView.Adapter<? extends VH> adapter) {
        this.adapter = adapter;
        return thisInstance;
    }

    public final B onLayoutManager(Consumer<RecyclerView.LayoutManager> layoutManagerConsumer) {
        this.layoutManagerConsumer = layoutManagerConsumer;
        return thisInstance;
    }

    public final B withLinearLayoutManager() {
        layoutManagerType = LINEAR_LAYOUT_MANAGER;
        return thisInstance;
    }

    public final B withGridLayoutManager(int spanCount) {
        layoutManagerType = GRID_LAYOUT_MANAGER;
        this.spanCount = spanCount;
        return thisInstance;
    }

    public final B withStaggeredGridLayoutManager(int spanCount) {
        layoutManagerType = STAGGERED_GRID_LAYOUT_MANAGER;
        this.spanCount = spanCount;
        return thisInstance;
    }

    public final B withCustomLayoutManager(RecyclerView.LayoutManager layoutManager) {
        this.customLayoutManager = layoutManager;
        return thisInstance;
    }

    public final B withRecycledViewPool(RecyclerView.RecycledViewPool recycledViewPool) {
        this.recycledViewPool = recycledViewPool;
        return thisInstance;
    }

    public final B withInconsistencyHandler(Consumer<IndexOutOfBoundsException> handler) {
        this.handler = handler;
        return thisInstance;
    }

    public final B withEndlessScrollCallback(int threshold, @NonNull Consumer<Integer> endlessScrollConsumer) {
        this.endlessScrollVisibleThreshold = threshold;
        this.endlessScrollConsumer = endlessScrollConsumer;
        return thisInstance;
    }

    public final B addStateListener(@NonNull Consumer<Integer> stateListener) {
        this.stateConsumers.add(stateListener);
        return thisInstance;
    }

    public final B addScrollListener(@NonNull BiConsumer<Integer, Integer> scrollListener) {
        this.displacementConsumers.add(scrollListener);
        return thisInstance;
    }

    public final B addDecoration(@NonNull RecyclerView.ItemDecoration decoration) {
        this.itemDecorations.add(decoration);
        return thisInstance;
    }

    public final B withRefreshLayout(@NonNull SwipeRefreshLayout refreshLayout, Runnable refreshAction) {
        this.refreshLayout = refreshLayout;
        refreshLayout.setOnRefreshListener(refreshAction::run);
        return thisInstance;
    }

    public final B withPlaceholder(@NonNull ListPlaceholder<T> placeholder) {
        this.placeholder = placeholder;
        return thisInstance;
    }

    public final B withSwipeDragOptions(@NonNull SwipeDragOptions<VH> swipeDragOptions) {
        this.swipeDragOptions = swipeDragOptions;
        return thisInstance;
    }

    public abstract S build();

    protected EndlessScroller buildEndlessScroller(RecyclerView.LayoutManager layoutManager) {
        return endlessScrollConsumer == null ? null : new EndlessScroller(endlessScrollVisibleThreshold, layoutManager) {
            @Override
            public void onLoadMore(int currentItemCount) {
                endlessScrollConsumer.accept(currentItemCount);
            }
        };
    }

    @CallSuper
    protected List<RecyclerView.OnScrollListener> buildScrollListeners() {
        int stateConsumersSize = stateConsumers.size();
        int scrollConsumersSize = displacementConsumers.size();
        int max = Math.max(stateConsumersSize, scrollConsumersSize);

        List<RecyclerView.OnScrollListener> scrollListeners = new ArrayList<>(max);

        for (int i = 0; i < max; i++) {
            final Consumer<Integer> consumer;
            final BiConsumer<Integer, Integer> biConsumer;

            if (i < stateConsumersSize) consumer = stateConsumers.get(i);
            else consumer = null;
            if (i < scrollConsumersSize) biConsumer = displacementConsumers.get(i);
            else biConsumer = null;

            scrollListeners.add(new RecyclerView.OnScrollListener() {
                @Override
                public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                    if (biConsumer != null) biConsumer.accept(dx, dy);
                }

                @Override
                public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                    if (consumer != null) consumer.accept(newState);
                }
            });
        }

        stateConsumers.clear();
        displacementConsumers.clear();
        return scrollListeners;
    }

    protected RecyclerView.LayoutManager buildLayoutManager() {
        RecyclerView.LayoutManager layoutManager;
        switch (layoutManagerType) {
            case STAGGERED_GRID_LAYOUT_MANAGER:
                layoutManager = new StaggeredGridLayoutManager(spanCount, StaggeredGridLayoutManager.VERTICAL) {
                    @Override
                    public void onLayoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
                        handleLayout(recycler, state, super::onLayoutChildren);
                    }
                };
                break;
            case GRID_LAYOUT_MANAGER:
                layoutManager = new GridLayoutManager(recyclerView.getContext(), spanCount) {
                    @Override
                    public void onLayoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
                        handleLayout(recycler, state, super::onLayoutChildren);
                    }
                };
                break;
            case LINEAR_LAYOUT_MANAGER:
                layoutManager = new LinearLayoutManager(recyclerView.getContext()) {
                    @Override
                    public void onLayoutChildren(RecyclerView.Recycler recycler, RecyclerView.State state) {
                        handleLayout(recycler, state, super::onLayoutChildren);
                    }
                };
                break;
            default:
                if (customLayoutManager != null) layoutManager = customLayoutManager;
                else throw new IllegalArgumentException("LayoutManager must be provided");
                break;
        }

        if (handler == null) Log.w(TAG, "InconsistencyHandler is not provided, " +
                "inconsistencies in the RecyclerView adapter will cause crashes at runtime");

        if (layoutManagerConsumer != null) layoutManagerConsumer.accept(layoutManager);

        return layoutManager;
    }

    private void handleLayout(RecyclerView.Recycler recycler, RecyclerView.State state,
                              BiConsumer<RecyclerView.Recycler, RecyclerView.State> function) {
        try {
            function.accept(recycler, state);
        } catch (IndexOutOfBoundsException e) {
            if (handler != null) handler.accept(e);
            else throw e;
        }
    }
}

File: recyclerview/src/main/java/com/tunjid/androidbootstrap/recyclerview/AbstractListManagerBuilder.kt
package com.tunjid.androidbootstrap.recyclerview

import android.util.Log
import androidx.annotation.CallSuper
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.RecyclerView.ViewHolder
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.tunjid.androidbootstrap.recyclerview.ListManager.Companion.TAG
import java.util.*
import kotlin.math.max

/**
 * Abstract thisInstance class for creating [scrollmanagers][ListManager]
 *
 *
 * The breakdown for the generic types are as follows:
 *
 *
 * B: The implicit type of the Builder. This is only necessary to make the return type for each
 * thisInstance method be the type of any inheriting class, and not this base class. Otherwise inheritors of
 * this subclass will need to override each method here to return their custom thisInstance type to
 * maintain the fluency of the API which is fairly tedious.
 *
 *
 * S: The type of the [ListManager] to be built.
 *
 *
 * VH: The [ViewHolder] type in the [RecyclerView]
 *
 *
 * T: The type bound in the [ListPlaceholder]
 */
@Suppress("unused")
abstract class AbstractListManagerBuilder<B : AbstractListManagerBuilder<B, S, VH, T>, S : ListManager<VH, T>, VH : ViewHolder, T> {

    protected var spanCount: Int = 0
    protected var layoutManagerType: Int = 0
    protected var endlessScrollVisibleThreshold: Int = 0
    protected var hasFixedSize: Boolean = false

    protected var placeholder: ListPlaceholder<T>? = null
    protected var refreshLayout: SwipeRefreshLayout? = null

    protected var recyclerView: RecyclerView? = null
    protected var adapter: RecyclerView.Adapter<out VH>? = null
    protected var customLayoutManager: RecyclerView.LayoutManager? = null

    protected var swipeDragOptions: SwipeDragOptions<VH>? = null
    protected var recycledViewPool: RecyclerView.RecycledViewPool? = null

    protected var endlessScrollConsumer: ((Int) -> Unit)? = null
    protected var layoutManagerConsumer: ((RecyclerView.LayoutManager) -> Unit)? = null
    protected var handler: ((IndexOutOfBoundsException) -> Unit)? = null

    protected var itemDecorations: MutableList<RecyclerView.ItemDecoration> = ArrayList()
    protected var stateConsumers: MutableList<(Int) -> Unit> = ArrayList()
    protected var displacementConsumers: MutableList<(Int, Int) -> Unit> = ArrayList()

    @Suppress("UNCHECKED_CAST")
    protected val thisInstance: B = this as B

    fun setHasFixedSize(): B = thisInstance.apply { this.hasFixedSize = true }

    fun withRecyclerView(recyclerView: RecyclerView): B = thisInstance.apply { this.recyclerView = recyclerView }

    fun withAdapter(adapter: RecyclerView.Adapter<out VH>): B = thisInstance.apply { this.adapter = adapter }

    fun onLayoutManager(layoutManagerConsumer: (RecyclerView.LayoutManager) -> Unit): B = thisInstance.apply { this.layoutManagerConsumer = layoutManagerConsumer }

    fun withLinearLayoutManager(): B = thisInstance.apply { layoutManagerType = LINEAR_LAYOUT_MANAGER }

    fun withGridLayoutManager(spanCount: Int): B {
        layoutManagerType = GRID_LAYOUT_MANAGER
        this.spanCount = spanCount
        return thisInstance
    }

    fun withStaggeredGridLayoutManager(spanCount: Int): B {
        layoutManagerType = STAGGERED_GRID_LAYOUT_MANAGER
        this.spanCount = spanCount
        return thisInstance
    }

    fun withCustomLayoutManager(layoutManager: RecyclerView.LayoutManager): B = thisInstance.apply { this.customLayoutManager = layoutManager }

    fun withRecycledViewPool(recycledViewPool: RecyclerView.RecycledViewPool): B = thisInstance.apply { this.recycledViewPool = recycledViewPool }

    fun withInconsistencyHandler(handler: (IndexOutOfBoundsException) -> Unit): B = thisInstance.apply { this.handler = handler }

    fun withEndlessScrollCallback(threshold: Int, endlessScrollConsumer: (Int) -> Unit): B {
        this.endlessScrollVisibleThreshold = threshold
        this.endlessScrollConsumer = endlessScrollConsumer
        return thisInstance
    }

    fun addStateListener(stateListener: (Int) -> Unit): B = thisInstance.apply { this.stateConsumers.add(stateListener) }

    fun addScrollListener(scrollListener: (Int, Int) -> Unit): B = thisInstance.apply { this.displacementConsumers.add(scrollListener) }

    fun addDecoration(decoration: RecyclerView.ItemDecoration): B = thisInstance.apply { this.itemDecorations.add(decoration) }

    fun withRefreshLayout(refreshLayout: SwipeRefreshLayout, refreshAction: () -> Unit): B {
        this.refreshLayout = refreshLayout
        refreshLayout.setOnRefreshListener { refreshAction() }
        return thisInstance
    }

    fun withPlaceholder(placeholder: ListPlaceholder<T>): B = thisInstance.apply { this.placeholder = placeholder }

    fun withSwipeDragOptions(swipeDragOptions: SwipeDragOptions<VH>): B = thisInstance.apply { this.swipeDragOptions = swipeDragOptions }

    abstract fun build(): S

    protected fun buildEndlessScroller(layoutManager: RecyclerView.LayoutManager): EndlessScroller? =
            if (endlessScrollConsumer == null) null
            else object : EndlessScroller(endlessScrollVisibleThreshold, layoutManager) {
                override fun onLoadMore(currentItemCount: Int) {
                    endlessScrollConsumer?.invoke(currentItemCount)
                }
            }

    @CallSuper
    protected fun buildScrollListeners(): List<RecyclerView.OnScrollListener> {
        val stateConsumersSize = stateConsumers.size
        val scrollConsumersSize = displacementConsumers.size
        val max = max(stateConsumersSize, scrollConsumersSize)

        val scrollListeners = ArrayList<RecyclerView.OnScrollListener>(max)

        for (i in 0 until max) {

            val consumer: ((Int) -> Unit)? =
                    if (i < stateConsumersSize) stateConsumers[i]
                    else null

            val biConsumer: ((Int, Int) -> Unit)? =
                    if (i < scrollConsumersSize) displacementConsumers[i]
                    else null

            scrollListeners.add(object : RecyclerView.OnScrollListener() {
                override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                    biConsumer?.invoke(dx, dy)
                }

                override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                    consumer?.invoke(newState)
                }
            })
        }

        stateConsumers.clear()
        displacementConsumers.clear()

        return scrollListeners
    }

    protected fun buildLayoutManager(): RecyclerView.LayoutManager {
        val context = recyclerView?.context
                ?: throw IllegalArgumentException("RecyclerView not provided")
        val layoutManager: RecyclerView.LayoutManager = when (layoutManagerType) {
            STAGGERED_GRID_LAYOUT_MANAGER -> object : StaggeredGridLayoutManager(spanCount, VERTICAL) {
                override fun onLayoutChildren(recycler: RecyclerView.Recycler?, state: RecyclerView.State) =
                        handleLayout { super.onLayoutChildren(recycler, state) }
            }
            GRID_LAYOUT_MANAGER -> object : GridLayoutManager(context, spanCount) {
                override fun onLayoutChildren(recycler: RecyclerView.Recycler?, state: RecyclerView.State) =
                        handleLayout { super.onLayoutChildren(recycler, state) }
            }
            LINEAR_LAYOUT_MANAGER -> object : LinearLayoutManager(context) {
                override fun onLayoutChildren(recycler: RecyclerView.Recycler?, state: RecyclerView.State) =
                        handleLayout { super.onLayoutChildren(recycler, state) }
            }
            else -> customLayoutManager
                    ?: throw IllegalArgumentException("LayoutManager must be provided")
        }

        if (handler == null)
            Log.w(TAG, "InconsistencyHandler is not provided, " + "inconsistencies in the RecyclerView adapter will cause crashes at runtime")

        layoutManagerConsumer?.invoke(layoutManager)

        return layoutManager
    }

    private fun RecyclerView.LayoutManager.handleLayout(superCall: () -> Unit) {
        try {
            superCall()
        } catch (e: IndexOutOfBoundsException) {
            handler?.invoke(e) ?: throw e
        }
    }

    companion object {

        private const val LINEAR_LAYOUT_MANAGER = 1
        private const val GRID_LAYOUT_MANAGER = 2
        private const val STAGGERED_GRID_LAYOUT_MANAGER = 3
    }
}
